HDDS-6244. ContainerBalancer metrics don't show updated values in JMX

siddhantsangwan opened the pull request:
## What changes were proposed in this pull request?

The `ContainerBalancerMetrics` class is responsible for recording metrics. These metrics always show 0 when accessed through JMX, while expected values are greater than 0. 

I deleted some metrics of the type `MutableGaugeLong` and instead used the type `MutableCounterLong`. Introduced a new metric `countIterations` that keeps count of the number of iterations that balancer has run for.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6244

## How was this patch tested?

Updated TestContainerBalancer#testMetrics()

siddhantsangwan commented:
@lokeshj1703 @JacksonYao287 please take a look!

lokeshj1703 reviewed (commented):
@siddhantsangwan Thanks for working on this! The changes look good to me. I have a few comments inline.

siddhantsangwan commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerMetrics.java:
```diff
@@ -37,27 +36,16 @@
-  private MutableGaugeLong dataSizeMovedGB;
+  @Metric(about = "Number of containers that Container Balancer moved" +
+      " in the last iteration.")
+  private MutableCounterLong movedContainersNum;
```
Sure, I'll update the names. Aggregate metrics - you mean total no. of containers moved across iterations etc? For that we'll need to save container ids across iterations to avoid counting the same container multiple times.

lokeshj1703 replied:
As we discussed we can have a simple metric. If a container is moved multiple times we can still count those moves. But this can be done in a separate jira.

siddhantsangwan commented:
@lokeshj1703 I've replaced the current metrics and included more testing.

JacksonYao287 reviewed (commented):
thanks @siddhantsangwan for the work ! the change looks good.
I also suggest that we can add aggregate metrics(total containers  and total datasize). but this can be implemented in a seperate jira.

lokeshj1703 commented on hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/balancer/TestContainerBalancer.java:
```diff
@@ -305,16 +302,14 @@ public void containerBalancerShouldObeyMaxSizeToMoveLimit() {
     // balancer should not have moved more size than the limit
     Assert.assertFalse(containerBalancer.getSizeMovedPerIteration() >
         10 * OzoneConsts.GB);
+    Assert.assertFalse(
+        containerBalancer.getMetrics().getDataSizeMovedGBInLatestIteration() >
+            10);
```
Could we also add assertion that metric is greater than 0?

lokeshj1703 commented on hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/balancer/TestContainerBalancer.java:
```diff
@@ -214,15 +215,12 @@ public void unBalancedNodesListShouldBeEmptyWhenClusterIsBalanced() {
     containerBalancer.stop();
+    ContainerBalancerMetrics metrics = containerBalancer.getMetrics();
     Assert.assertEquals(0, containerBalancer.getUnBalancedNodes().size());
+    Assert.assertEquals(0, metrics.getDatanodesNumUnbalanced());
```
Can we also assert that unbalanced datanodes and size is some expected number in a simulated test, if it is not very difficult to add?

siddhantsangwan replied:
We're already testing this in `testMetrics()` at line 513:
```
    Assert.assertEquals(determineExpectedUnBalancedNodes(
            balancerConfiguration.getThreshold()).size(),
        metrics.getDatanodesNumUnbalanced());
```

lokeshj1703 reviewed (commented):
Thanks @siddhantsangwan for updating the PR! I have some minor comments.

siddhantsangwan commented:
Thanks for reviewing @lokeshj1703 @JacksonYao287 
Addressed your comments and opened a new jira for aggregate metrics: https://issues.apache.org/jira/browse/HDDS-6362

lokeshj1703 commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerMetrics.java:
```diff
@@ -45,14 +45,14 @@
+  private MutableCounterLong dataSizeUnbalancedGB;
 
   @Metric(about = "Number of unbalanced datanodes.")
   private MutableCounterLong datanodesNumUnbalanced;
```
I will also recommend changing names of other metrics. Like datanodesNum -> numDatanodes and movedContainersNum -> numMovedContainers.
We can do it in a separate PR as well.

siddhantsangwan replied:
Oh, I remember we had decided on this naming convention in a previous PR: https://github.com/apache/ozone/pull/2230#discussion_r637872787

But this does conflict with `numIterations` where "num" is the prefix.

lokeshj1703 replied:
I see. My idea is that it should be easy to search these metrics in prometheus. Maybe we need to have common prefix for balancer then. Can we check some naming guide here?

siddhantsangwan replied:
I looked into some examples. We've used "num" as a prefix in other places, such as `numCacheHits`, `numCacheMisses`, `numInFlightReplications`, `numWriteStateMachineOps` etc.

siddhantsangwan replied:
@lokeshj1703 I've fixed the naming.

lokeshj1703 reviewed (approved):
Thanks for updating the PR! The changes look good to me. +1.
I have one minor comment, let me know if you would like to address it in this or new PR.

lokeshj1703 commented:
@siddhantsangwan Thanks for the contribution! @JacksonYao287 Thanks for the reviews! I have committed the PR to master branch.
